File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> stones, int k) {
    int answer = 0;
    // 모든 조약돌은 전체적으로 n을 뺐을때 0이하가 k번 연속으로 발생하는지 확인
    int end = *max_element(stones.begin(), stones.end()); 
    int start = 0;
    while(start <= end){ // 10^3
        int mid = (start+end)/2; 
        int jump = 0;
        bool avail = true;
        for(int i = 0; i < stones.size(); i++){ // 200,000
            if(stones[i] - mid < 0){
                jump++;
                if(jump >= k) {
                    avail = false; break;
                }
            }else{
                jump = 0;
            }
        }
        if(!avail) {
            end = mid-1;
        }else{
            start = mid+1;
            answer = mid;
        }
    }
    
    //  최대 몇 명까지 징검다리를 건널 수 있는지 return
    return answer ;
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.*;

class Solution {
    
    
    public int solution(int[] stones, int k) {
        int answer = 0;
        
        // 이분탐색으로 명수 구하기
        int left = 1; int right = 200000000;
        while(left <= right){
            int mid = (left + right)/2; 
            
            int seq = 0; boolean avail = true;
            for(int s : stones){
                if(s < mid) {
                    seq++;
                    if(seq +1 > k){
                        avail = false; break;
                    }
                }
                else seq = 0;
            }
            if(!avail) right = mid-1; 
            else {
                answer = mid;
                left = mid +1;
            }
        }
        
        // 최대 몇 명까지 징검다리를 건널 수 있는지 return
        return answer;
    }
}